go/mcap/indexed_message_iterator.go

```diff
@@ -286,28 +432,85 @@ func (it *indexedMessageIterator) Next(_ []byte) (*Schema, *Channel, *Message, e
-			if err != nil {
+	if !it.useHeap {
+		if it.curMessageIndex >= len(it.messageIndexes) {
+			if it.curChunkIndex >= len(it.chunkIndexes) {
+				return nil, nil, nil, io.EOF
+			}
+			chunkIndex := it.chunkIndexes[it.curChunkIndex]
+			if err := it.loadChunk(chunkIndex); err != nil {
 				return nil, nil, nil, err
 			}
-			continue
+			it.curMessageIndex = 0
+			it.curChunkIndex++
+			return it.Next2(msg)
 		}
-		chunkOffset := ri.messageIndexEntry.Offset
-		length := binary.LittleEndian.Uint64(ri.buf[chunkOffset+1:])
-		messageData := ri.buf[chunkOffset+1+8 : chunkOffset+1+8+length]
-		message, err := ParseMessage(messageData)
-		if err != nil {
+		messageIndex := it.messageIndexes[it.curMessageIndex]
+		decompressedChunk := &it.decompressedChunks[0]
+		if err := loadMessageAtOffset(decompressedChunk.buf, messageIndex.Offset, msg); err != nil {
 			return nil, nil, nil, err
 		}
-		channel := it.channels[message.ChannelID]
-		schema := it.schemas[channel.SchemaID]
-		return schema, channel, message, nil
+		decompressedChunk.unreadMessages--
+		it.curMessageIndex++
+		schema, channel, err := it.getSchemaAndChannel(msg.ChannelID)
+		return schema, channel, msg, err
+	}
+
+	if it.indexHeap.len() == 0 {
+		return nil, nil, nil, io.EOF
 	}
-	return nil, nil, nil, io.EOF
+	ri, err := it.indexHeap.Pop()
+	if err != nil {
+		return nil, nil, nil, err
+	}
+	if ri.ChunkSlotIndex == -1 {
+		if err := it.loadChunk(ri.chunkIndex); err != nil {
+			return nil, nil, nil, err
+		}
+		return it.Next2(msg)
+	}
+	decompressedChunk := &it.decompressedChunks[ri.ChunkSlotIndex]
+	if err := loadMessageAtOffset(decompressedChunk.buf, ri.MessageOffsetInChunk, msg); err != nil {
+		return nil, nil, nil, err
+	}
+	decompressedChunk.unreadMessages--
+	schema, channel, err := it.getSchemaAndChannel(msg.ChannelID)
+	return schema, channel, msg, err
+}
+
+func (it *indexedMessageIterator) getSchemaAndChannel(channelID uint16) (*Schema, *Channel, error) {
+	channel := slicemap.GetAt(it.channels, channelID)
+	if channel == nil {
+		return nil, nil, fmt.Errorf("message with unrecognized channel ID %d", channelID)
+	}
+	schema := slicemap.GetAt(it.schemas, channel.SchemaID)
+	if schema == nil && channel.SchemaID != 0 {
+		return nil, nil, fmt.Errorf("channel %d with unrecognized schema ID %d", channelID, channel.SchemaID)
+	}
+	return schema, channel, nil
+}
+
+// loadMessageAtOffset loads the Message record from `decompressedChunk` into `msg`.
+func loadMessageAtOffset(decompressedChunk []byte, offset uint64, msg *Message) error {
+	length := binary.LittleEndian.Uint64(decompressedChunk[offset+1:])
+	messageData := decompressedChunk[offset+1+8 : offset+1+8+length]
+	previousData := msg.Data
+	if err := msg.PopulateFrom(messageData); err != nil {
+		return err
+	}
+	// after PopulateFrom, msg.Data is a slice of `decompressedChunk`. Copy it out so
+	// that the `decompressedChunk` buffer can be reused later.
+	msg.Data = append(previousData[:0], msg.Data...)
```

Comment:
note that if PopulateFrom performs the copy into the existing msg.Data, we can drop the copy here